### PR TITLE
[Exclusivity] Sink may release release instructions out of access scopes

### PR DESCRIPTION
**Pull request by shajrawi**
rdar://problem/44121388

General case:
begin_access A
...
strong_release / release_value / destroy
end_access

The release instruction can be sunk below the end_access instruction,
This extends the lifetime of the released value, but, might allow us to
Mark the access scope as no nested conflict.

### CODE REVIEW

**Comment by shajrawi** [comment]
@swift-ci Please test

**Comment by shajrawi** [comment]
@atrick , @eeckstein can either one of you please review?

**Review by eeckstein — APPROVED** [review]
lgtm!

File: lib/SILOptimizer/Transforms/AccessEnforcementReleaseSinking.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                              << *bottomEndAccessInst << "\n");
+      // We need to avoid iterator invalidation:
+      // We know this is not the last instruction of the block:
+      // 1) not a TermInst
+      // 2) bottomEndAccessInst != nil
+      assert(reverseIt != block.rbegin() &&
+             "Did not expect a sinkable instruction at block's end");
+      // Go back to previous iteration
+      auto prevIt = reverseIt;
+      --prevIt;
```
Comment (atrick) [inline lib/SILOptimizer/Transforms/AccessEnforcementReleaseSinking.cpp]: [nit] `auto prevIt = std::prev(reverseIt)`

**Review by atrick — COMMENTED** [review]
Nice!

**Pull request merged**